Initial scope implementation.

kbsriram opened the pull request:
This is an initial implementation to support scopes for the
Java language bindings. It currently provides only hierarchical
name scopes.

Please see https://github.com/tensorflow/tensorflow/issues/7149
to track the master issue.

kbsriram committed:
Initial scope implementation.

This is an initial implementation to support scopes for the
Java language bindings. It currently provides only hierarchical
name scopes.

Please see https://github.com/tensorflow/tensorflow/issues/7149
to track the master issue.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

kbsriram commented:
@Moriadry thanks! I've started [an informal group](https://groups.google.com/forum/#!forum/tensorflow-java-dev-unofficial) to find out who's working on what in this particular area, and consolidate efforts from interested people - we can continue the conversation there.

Moriadry commented:
@kbsriram Got it, I already submit a request to join, many thanks!

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Scope.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    private Scope build() {
+      return new Scope(this);
```
I'm a little confused by this Scope, Scope.Builder, NameScope, NameScope.Builder relationships. It seems like a lot of abstraction that can be simplified, but perhaps I'm mistaken.

I was imagining a single class, `Scope`, with something like this:

```
public final class Scope {
   ...
   // All the state needed for a Scope
   private final Graph graph;
   private final String name;  // "fully qualified name", i.e., path from root to here with slashes
   private final Map<String, int> opNames;  // Map from op-name (ignoring scope name prefix) to count
   private final Map<String, int> subNames;  // Map from sub-scope name to count
  
   // Unfortunate additional state needed for withOpName.
   // null to use the operation name as the node name
   private String opName;
}
```

And then:

- `new Scope` would set all fields (`name` to `""`, `opName` to `null`, the maps to newly allocated maps)
- `withSubScope` would do the name as `new Scope` except set `name` to `this.name + "/" + name`
- `withOpName` would copy all the fields of the "parent" `Scope` (so that the count maps are shared), but set `opName` differently

I _think_ doing so will make the scope code easier to follow as there will be fewer layers to navigate through.
Thoughts?

kbsriram replied:
Thank you for the thoughtful and detailed comments @asimshankar - much appreciated! I'll update this PR to reflect your suggestions.

May we sound you out a bit further on `Scope` and `NameScope`? No problem collapsing them in one class, though this what we were thinking.

There are two classes now, with `Scope` delegating to `NameScope`. We organized it as a main class delegating to a specialized class so we can add other self-contained groups of properties like "control dependencies" without putting everything in one implementation.

In this approach, the public `Scope` would be augmented to hold package private `NameScope` and `ControlDependency` objects, delegating to each as appropriate.

A `Scope` carrying a certain name hierarchy would then implement `withControlDependencies()` by reusing the existing `NameScope` and just creating a new `ControlDependency` object. `withName()` and `withOpName()` would do the opposite, etc. Having specialized classes to manage each group of properties would also make it a bit easier to test each in isolation.

What do you think?

[Re: Builders to construct classes rather than raw constructors; we were following [https://engdoc.corp.google.com/eng/doc/devguide/java/practices/builders.md?cl=head](https://engdoc.corp.google.com/eng/doc/devguide/java/practices/builders.md?cl=head) style that  prefers Builders or a static factory method to raw constructors. We chose Builders here as `NameScope` manages 3/4 constructor parameters of interest, and `Scope` currently has just one but we anticipate at least two after adding `ControlDependency`.  That said, we're happy to switch to raw constructors if you prefer this style for Tensorflow, we can switch 
 in this and subsequent PRs, please let us know.]

asimshankar replied:
Appreciate the thoughtful reply. A few replies:

- RE: Builders: I took those guidelines to mean that that builders are great as a replacement for factory methods when there are many parameters, or optional parameters or when calling code is less self-documenting that it should be. However, in this particular case, the proposal is to have only a single constructor (`new Scope(graph)`) and then named methods for child scopes (`withFoo()`), hence I don't think the preconditions for those guidelines are met. The builder in this PR is private and used for the implementation and IMHO in this case they are making the implementation harder to follow. So, I'd like to lean against the use of the private builders here. Sound fair?

- RE: Helper scopes: I see. IIUC, you're trying to encapsulate all the state needed for "names" in one class, all the state needed for "control dependencies" in another class etc. So, instead of having each `withFoo()` method worry about how it handles every private state member of the `Scope` class, it can deal with groups? I can buy that - let's see how things look once the builders are gone (I suspect the implementation of both classes will become much smaller) and see how it goes then.

kbsriram replied:
That sounds great :-) Your summary about helper scopes is just how we're thinking about it as well. I'll update the PR with all the changes you suggested (but keep the helper scopes as-is) and go from there. Thanks!

asimshankar commented on tensorflow/java/src/test/java/org/tensorflow/ScopeTest.java:
```diff
@@ -0,0 +1,184 @@
+      Const c2 = Const.create(s, 7);
+      assertEquals("Const_1", c2.input().op().name());
+      Const c3 = Const.create(s.withOpName("four"), 4);
+      assertEquals("four", c3.input().op().name());
```
Perhaps also add, for documentation purposes, test for the following scenario (a second constant with the same name):

```java
Const c4 = Const.create(s.withOpName("four"), 4);
```

asimshankar commented on tensorflow/java/src/test/java/org/tensorflow/ScopeTest.java:
```diff
@@ -0,0 +1,184 @@
+      assertEquals("Const", c1.input().op().name());
+      Const c2 = Const.create(s, 7);
+      assertEquals("Const_1", c2.input().op().name());
+      Const c3 = Const.create(s.withOpName("four"), 4);
```
What do you think about `withName` instead of `withOpName`? So, a client will deal with `s.withSubScope("foo")` and `s.withName("bar")`?

asimshankar commented on tensorflow/java/src/test/java/org/tensorflow/ScopeTest.java:
```diff
@@ -0,0 +1,184 @@
+
+  // Convenience interface
+  // TODO: replace when standardized
+  interface Input {
```
For now, I know we might undo this, but might I recommend getting rid of this interface and the `input()` method in the "handwritten" classes? Without the complete context, things like `public Output input() { return output; }` and the use in `Const(...).input().op()...` seem confusing. Better to keep this patch focused on the classes it adds?

asimshankar reviewed (commented):
Thanks for this PR and sorry for the late reply! Somehow I missed that I hadn't sent out these comments and just noticed. 

My main concern is about the implementation complexity of `Scope` with many helper classes. I was wondering if we can simplify the implementation down to a single class with a small number of "state" member variables.

kbsriram committed:
Modifications per recommended changes

* moved classes to org.tensorflow.op package
* BUILD changes to keep op package in a separate target
* Updated ci_build to also build new target
* Replaced static factory method for Scope with public raw constructor
* Replaced Builders for NameScope and Scope with raw constructors
* Addressed JavaDoc nits
* Added comments pointing to reasons for operator name restrictions
* Merged NameScopeTest into ScopeTest
* renamed scope.withOpName -> scope.withName

asimshankar commented on tensorflow/java/BUILD:
```diff
@@ -15,6 +15,16 @@ java_library(
 
+# Operator convenience wrappers for the core tensorflow library.
+java_library(
+    name = "tensorflow_op",
```
Instead of a separate `java_library`, how about including `:java_op_sources` in the `tensorflow` target? That way we end up with one `.jar`. Otherwise we have to worry about distribution of this jar and will have to deal with more complicated instructions on www.tensorflow.org/install/install_java

(The Android builds use the `java_sources` filegroup so they won't include these files).

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+ * <p>This class is package private, user code creates {@link Scope} which internally delegates
+ * calls to an underlying {@code NameScope}.
+ *
+ * <p>This class is thread-safe.
```
Does it need to be? Do we want to encourage graph construction in multiple threads?
I'm tempted to suggest that we:

- Change this to `<p>Objects of this class are NOT thread-save`
- Remove the use of `synchronized` in the implementation

It will be easier to make this thread-safe in the future if needed, but harder to remove the locks if they are a performance bottleneck. Or perhaps I'm overthinking this - it's just a single lock :) Thoughts?

kbsriram replied:
If graph construction time seems a non-issue, we can close off the option entirely as you suggest.

One thought: should you be considering more of a wait-and-see approach - it may be less disruptive to improve a less performant but correct implementation than change it from "not thread-safe" to "thread-safe." The reason is that user-side code will likely need to restructure itself to take advantage in the latter case; but in the former case it will improve directly as the framework code improves.

For what it's worth, the locks here are per scope-instance; so to the extent different threads use different scopes there should be no contention. Aiui, possibly the place to watch out may be  [the per-graph global lock](https://github.com/tensorflow/tensorflow/blob/ce02c770fb269e5e607da459bde4f580ef108137/tensorflow/c/c_api_internal.h#L67) in the underlying c_api abstraction that guards operation creation itself.

asimshankar replied:
I was thinking that making the API not promise thread-safety and bolting it on later will be less disruptive (as any correct usage of the API will not run into trouble) than the other way around.

Compounded with the fact that I don't quite see a pressing need for multi-threaded graph construction. Do you see any?

(I agree that the lock overhead itself may not be a big deal, but overall I'm trying to err on the side of simplicity instead of a feature for a use-case that we don't foresee)

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+
+  NameScope withSubScope(String scopeName) {
+    if (baseName == null) {
+      checkPattern(OP_NAME_REGEX, scopeName);
```
Optional: I find this easier to read:

```
checkPattern(baseName == null ? OP_NAME_REGEX : SUBSCOPE_NAME_REGEX, scopeName);
```

but maybe it's just me. Your call.

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+  // LETTER_DIGIT_DOT, followed by zero or more LETTER_DIGIT_DASH_DOT_SLASH_UNDERSCORE. SLASH is not
+  // permitted in actual user-supplied names to NameScope - it is used as a reserved character to
+  // separate subcomponents within fully qualified names.
+  private static final Pattern OP_NAME_REGEX = Pattern.compile("[A-Za-z0-9.][A-Za-z0-9_.\\-]*");
```
Isn't this supposed to be a slash (`/`) and not a backslash (`\`)? So: `/` instead of `\\`?

asimshankar replied:
Doh! Sorry, nevermind - misread the escaping :)

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+  // permitted in actual user-supplied names to NameScope - it is used as a reserved character to
+  // separate subcomponents within fully qualified names.
+  private static final Pattern OP_NAME_REGEX = Pattern.compile("[A-Za-z0-9.][A-Za-z0-9_.\\-]*");
+  private static final Pattern SUBSCOPE_NAME_REGEX = Pattern.compile("[A-Za-z0-9_.\\-]+");
```
Doesn't this regex make a single slash a valid subscope name. Should that be the case?
Actually, I'm not sure I understand why we need two regexps

asimshankar replied:
Doh! Sorry, nevermind - misread the escaping :)

asimshankar replied:
This makes `"-"` s valid subscope name (`withSubScope("-")`). Is that intentional? (Wondering if we can use a single regexp for now.

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+  //
+  // These two patterns are used to ensure fully qualified names always start with a
+  // LETTER_DIGIT_DOT, followed by zero or more LETTER_DIGIT_DASH_DOT_SLASH_UNDERSCORE. SLASH is not
+  // permitted in actual user-supplied names to NameScope - it is used as a reserved character to
```
IIUC, Python allows slashes in user-supplied names and this regexp also makes that possible. Why the comment?

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+    if (ids != null) {
+      this.ids = ids;
+    } else {
+      this.ids = new HashMap<String, Integer>();
```
I just realized that if we want to allow users to name ops with slashes in them (as is possible in Python), then we'd probably need the name-map attached to the root scope and not create new hash-maps per sub-scope - i.e., a single map that covers all the nodes in the graph. I believe that is the case in the C++ scope as well.

Should we replicate that here?

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+
+    if (baseOpName != null) {
+      // Use the base name instead to derive the subscope.
+      scopeName = baseOpName;
```
This is a bit confusing in that `withSubScope` ignores the argument if `baseOpName` is `null`.
One alternative would be that `withSubScope` ignores the `baseOpName` and just creates a new sub-scope.
So something like:

```java
Const.create(s.withName("foo"), ...) // Results in "foo"
Const.create(s.withName("foo").withSubScope("foo"), ...)  // Results in "foo/Const" - as if the withName was ignored
```

That way it is clearer than `withName` only applies to the name (suffix) of the op and doesn't affect the scoping prefixes ever.

What do you think?

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,158 @@
+
+  // If baseOpName is non-null, it is used to derive the unique name
+  // for operators rather than the provided default name.
+  private final String baseOpName;
```
How about `scopeName` and `opName` instead of `baseName` and `baseOpName`? Or perhaps `opPrefix` and `opName` - since one is intended to be a prefix for any ops created?

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/Scope.java:
```diff
@@ -0,0 +1,170 @@
+ * public class Constant {
+ *   public static Constant create(Scope scope, ...) {
+ *      scope.graph().opBuilder(
+ *        "Constant", scope.makeOpName("Constant"))
```
To be faithful to the real op name, `Const` instead of `Constant`?

asimshankar commented on tensorflow/java/src/test/java/org/tensorflow/op/ScopeTest.java:
```diff
@@ -0,0 +1,265 @@
+      output = o;
+    }
+
+    public Output output() {
```
Should we remove these `private` and `public` method qualifiers from the sample ops? (They are only needed for the tests and since they are inner classes, the test code is able to call them anyway (e.g., `Const.create` is private, but still callable - so the distinction between `private` and `public` here is immaterial anyway?)

Alternatively, make the `create` factory function `public` to be closer to where we want to head anyway?

kbsriram committed:
Modifications per recommended changes.

* Document Scope as not offering thread-safety
* Remove synchronized sections within Scope
* Restrict withName() and withSubScope() to only allow names matching
  [A-Za-z0-9.][A-Za-z0-9_.\\-]*
* Consolidate tensorflow and tensorflow_op into one BUILD target
* Fix mistake in opname in Javadoc example
* Change method qualifier in inner test classes to default
* rename baseName -> opPrefix, baseOpName -> opName

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,148 @@
+
+  NameScope withSubScope(String scopeName) {
+    checkPattern(NAME_REGEX, scopeName);
+
```
Out of curiosity, why all the blank lines in these short functions?
Maybe this is my C/Go background talking, but compact is better in these cases? :)

(Feel free to leave as is or change, your call - I was just curious)

kbsriram replied:
Wasn't a conscious choice :-) realize I habitually set off single-line comments with a blank line, and here probably that combined with removing unneeded blank lines. Out they go.

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,148 @@
+      throw new IllegalArgumentException("Names cannot be null");
+    }
+    if (!pattern.matcher(name).matches()) {
+      throw new IllegalArgumentException(String.format("Invalid name '%s'", name));
```
Let's add some more detail to this error, perhaps something like:

"invalid name: '%s' does not match the regular expression NAME_REGEX"

asimshankar reviewed (changes requested):
Thanks for your patience, excited to have this in.

One minor comment about the error message - other than that I think this is good to go.

kbsriram committed:
Small list of cleanups

* Better error message for invalid names
* Removed unneeded blank lines in short function

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java:
```diff
@@ -0,0 +1,147 @@
+      throw new IllegalArgumentException(
+          String.format(
+              "invalid name: '%s' does not match the regular expression %s",
+              name, NAME_REGEX_STRING));
```
I don't think you need NAME_REGEX_STRING, doesn't `NAME_REGEX.toString()` work just as well?

kbsriram replied:
Sure, that works too - ok to use pattern() instead, just to ensure it's clear what's being returned without referring to the documentation?

asimshankar replied:
Sure.

asimshankar reviewed (approved):
One comment about NAME_REGEX_STRING, otherwise LGTM

kbsriram committed:
Replace string constant in error message

Use Pattern.pattern() in error message rather than a
separately declared String constant.

kbsriram commented:
Gentle ping - could a member wake up the test this please ci bot?

martinwicke commented:
Jenkins, test this please!

asimshankar commented:
The single test failure (`//tensorflow/python:basic_session_run_hooks_test`) looks unrelated.

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Broken Mac test rebuilt here: http://ci.tensorflow.org/job/tensorflow-pull-requests-mac/5318/console

martinwicke commented:
Broken Makefile test rebuilt here: http://ci.tensorflow.org/job/tensorflow-pull-requests-makefile/9172/console

martinwicke commented:
Broken Android test rebuilt here: http://ci.tensorflow.org/job/tensorflow-pull-requests-android/5321/console
